### PR TITLE
Optionally allow only one active session per identity 

### LINKED ISSUE #139: Optionally allow only one active session per identity
**Is your feature request related to a problem? Please describe.**

In some cases, it is considered good practice to have only one simoultaneous logon session. This may apply to online banking and other highly sensitive services.

**Describe the solution you'd like**

It should be possible to deactivate all active logon session for that particular user on a successful login.

This could easily be achieved with the workflow engine (just write a workflow that removes all active sessions for the user).

**Context**

It should be made clear that this behavior may cause issues when trying to save data (e.g. submit form on a now logged out session). This should be used with care and consideration.

### PR DESCRIPTION
**Pull request by evalsocket**
## Related issue
This PR will close #139 

## Proposed changes

- [x] Add session_destroyer hook 
- [x] Add session_destroyer hook testcase
- [x] Create a new key for hook `	KeySessionDestroyer = "revoke_active_sesions" `
- [x] Register hook in `registry_default_hooks.go `

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation within the code base (if appropriate)
- [x] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

File: selfservice/hook/session_destroyer.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	return &SessionDestroyer{r: r}
+}
+
+func (e *SessionDestroyer) ExecuteRegistrationPostHook(w http.ResponseWriter, r *http.Request, a *registration.Request, s *session.Session) error {
+
+	return nil
+}
+
+func (e *SessionDestroyer) ExecuteLoginPostHook(w http.ResponseWriter, r *http.Request, a *login.Request, s *session.Session) error {
+	if err := e.r.SessionPersister().DeleteSession(r.Context(), s.Identity.ID); err != nil {
```
Comment (aeneasr) [inline selfservice/hook/session_destroyer.go]: `DeleteSession` accepts the session ID, so you can not use `s.Identity.ID` here. Instead, you need to add a new method to [`session.Persister`](https://github.com/ory/kratos/blob/master/session/persistence.go#L31) such as `DeleteSessionsFor(ctx context.Context, identityID uuid.UUID) error`.

Next, you need to implement a test for that function [around here](https://github.com/ory/kratos/blob/master/session/persistence.go#L67).

Next, you need to implement the actual storage logic [here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_session.go#L22).

Once that is done, you should run:

```
go test -short -tags sqlite ./...
```

in the root directory of the project and see if tests pass.

Then, you will use that new method you implemented here.

File: selfservice/hook/session_destroyer_test.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/ory/kratos/driver/configuration"
+	"github.com/ory/kratos/identity"
+	"github.com/ory/kratos/internal"
+	"github.com/ory/kratos/selfservice/hook"
+	"github.com/ory/kratos/session"
+	"github.com/ory/kratos/x"
+)
+
+func TestSessionIssuer(t *testing.T) {
```
Comment (aeneasr) [inline selfservice/hook/session_destroyer_test.go]: This can not work because the function name is already reserved by another test. This should be something like `TestSessionDestroyer`

File: driver/registry_default_hooks.go [hunk 3]
```diff
@@ -13,7 +13,6 @@ import (
 
 func (m *RegistryDefault) hooksPost(credentialsType identity.CredentialsType, configs []configuration.SelfServiceHook) postHooks {
 	var i postHooks
-
```
Comment (aeneasr) [inline driver/registry_default_hooks.go]: Revert this change

File: selfservice/hook/session_destroyer.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+type SessionDestroyer struct {
+	r sessionDestroyerDependencies
+}
+
+func NewSessionDestroyer(r sessionDestroyerDependencies) *SessionDestroyer {
+	return &SessionDestroyer{r: r}
+}
+
+func (e *SessionDestroyer) ExecuteRegistrationPostHook(w http.ResponseWriter, r *http.Request, a *registration.Request, s *session.Session) error {
+
```
Comment (aeneasr) [inline selfservice/hook/session_destroyer.go]: remove newline

File: selfservice/hook/session_destroyer_test.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	_, reg := internal.NewRegistryDefault(t)
+	viper.Set(configuration.ViperKeyURLsSelfPublic, "http://localhost/")
+	viper.Set(configuration.ViperKeyDefaultIdentityTraitsSchemaURL, "file://./stub/stub.schema.json")
+
+	var r http.Request
+	h := hook.NewSessionIssuer(reg)
+
+	t.Run("method=sign-in", func(t *testing.T) {
```
Comment (aeneasr) [inline selfservice/hook/session_destroyer_test.go]: This should be `method=ExecuteLoginPostHook`

File: selfservice/hook/session_destroyer_test.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+		w := httptest.NewRecorder()
+		sid := x.NewUUID()
+
+		i := identity.NewIdentity("")
+		require.NoError(t, reg.IdentityPool().CreateIdentity(context.Background(), i))
+		require.NoError(t, h.ExecuteLoginPostHook(w, &r, nil, &session.Session{ID: sid, Identity: i}))
```
Comment (aeneasr) [inline selfservice/hook/session_destroyer_test.go]: This test isn't actually testing the functionality of what you've implemented. Instead, you need to do something along the lines of:

1. Create a new identity
2. Create 2 or more sessions for that identity
3. Run `ExecuteLoginPostHook`
4. Check if these 2 (or more) sessions have been removed from the store. To prove that they're gone, `GetSession` should return an error of type `sqlcon.ErrNoRows`.

File: driver/registry_default_hooks.go (hunk 3 above)
Comment (evalsocket) [reply driver/registry_default_hooks.go]: done

File: selfservice/hook/session_destroyer.go (hunk 1 above)
Comment (evalsocket) [reply selfservice/hook/session_destroyer.go]: done

File: selfservice/hook/session_destroyer.go (hunk 4 above)
Comment (evalsocket) [reply selfservice/hook/session_destroyer.go]: done

File: selfservice/hook/session_destroyer_test.go (hunk 5 above)
Comment (evalsocket) [reply selfservice/hook/session_destroyer_test.go]: done

File: docs/.kratos.yaml [hunk 7]
```diff
@@ -68,12 +68,14 @@ selfservice:
     after:
       password:
         - run: session
+        - run: revoke_active_sesions
```
Comment (aeneasr) [inline docs/.kratos.yaml]: Typo, this should be `revoke_active_sessions`

File: docs/.kratos.yaml [hunk 8]
```diff
@@ -68,12 +68,14 @@ selfservice:
...
         - run: redirect
           config:
             default_redirect_url: http://test.kratos.ory.sh:4000/
             allow_user_defined_redirect: true
       oidc:
         - run: session
+        - run: revoke_active_sesions
```
Comment (aeneasr) [inline docs/.kratos.yaml]: See above

File: persistence/sql/persister_session.go [hunk 9]
```diff
@@ -27,3 +28,14 @@ func (p *Persister) CreateSession(ctx context.Context, s *session.Session) error
 func (p *Persister) DeleteSession(ctx context.Context, sid uuid.UUID) error {
 	return p.c.Destroy(&session.Session{ID: sid}) // This must not be eager or identities will be created / updated
 }
+
+func (p *Persister) DeleteSessionsFor(ctx context.Context, sid uuid.UUID) error {
+	count, err := p.c.RawQuery("DELETE FROM sessions WHERE identity_id =?", sid).ExecWithCount()
+	if err != nil {
+		return sqlcon.HandleError(err)
+	}
+	if count == 0 {
```
Comment (aeneasr) [inline persistence/sql/persister_session.go]: We don't really care if count == 0

File: persistence/sql/persister_session.go [hunk 10]
```diff
@@ -27,3 +28,14 @@ func (p *Persister) CreateSession(ctx context.Context, s *session.Session) error
 func (p *Persister) DeleteSession(ctx context.Context, sid uuid.UUID) error {
 	return p.c.Destroy(&session.Session{ID: sid}) // This must not be eager or identities will be created / updated
 }
+
+func (p *Persister) DeleteSessionsFor(ctx context.Context, sid uuid.UUID) error {
+	count, err := p.c.RawQuery("DELETE FROM sessions WHERE identity_id =?", sid).ExecWithCount()
```
Comment (aeneasr) [inline persistence/sql/persister_session.go]: We don't need `ExecWithCount`, `Exec` should be enough

File: selfservice/hook/session_destroyer_test.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+	t.Run("method=ExecuteLoginPostHook", func(t *testing.T) {
+		w := httptest.NewRecorder()
+		sid := x.NewUUID()
+
+		i := identity.NewIdentity("")
+		require.NoError(t, reg.IdentityPool().CreateIdentity(context.Background(), i))
+		err := reg.SessionPersister().CreateSession(context.Background(), &session.Session{ID: sid, Identity: i}))
+		require.NoError(t, err)
+		err = reg.SessionPersister().CreateSession(context.Background(), &session.Session{ID: sid, Identity: i}))
```
Comment (aeneasr) [inline selfservice/hook/session_destroyer_test.go]: This creates the same session twice, instead, it would be better to create two individual sessions with individual IDs

File: session/persistence.go [hunk 12]
```diff
@@ -75,5 +78,17 @@ func TestPersister(p interface {
 			_, err := p.GetSession(context.Background(), expected.ID)
 			require.Error(t, err)
 		})
+
+		t.Run("case=delete session for", func(t *testing.T) {
+			var expected Session
+			require.NoError(t, faker.FakeData(&expected))
+			require.NoError(t, p.CreateIdentity(context.Background(), expected.Identity))
+			require.NoError(t, p.CreateSession(context.Background(), &expected))
+			require.NoError(t, p.CreateSession(context.Background(), &expected))
```
Comment (aeneasr) [inline session/persistence.go]: Here too, it would be better to create two individual sessions and create them :) That ensures that really ALL sessions are revoked, not just one

File: selfservice/hook/session_destroyer_test.go (hunk 11 above)
Comment (evalsocket) [reply selfservice/hook/session_destroyer_test.go]: are you saying that i need two sid ?
```
err := reg.SessionPersister().CreateSession(context.Background(), &session.Session{ID: sid, Identity: i}))
err := reg.SessionPersister().CreateSession(context.Background(), &session.Session{ID: sid2, Identity: i}))
```

File: persistence/sql/persister_session.go (hunk 10 above)
Comment (evalsocket) [reply persistence/sql/persister_session.go]: done

File: persistence/sql/persister_session.go (hunk 9 above)
Comment (evalsocket) [reply persistence/sql/persister_session.go]: done

File: docs/.kratos.yaml (hunk 8 above)
Comment (evalsocket) [reply docs/.kratos.yaml]: done

File: docs/.kratos.yaml (hunk 7 above)
Comment (evalsocket) [reply docs/.kratos.yaml]: done

File: selfservice/hook/session_destroyer_test.go (hunk 11 above)
Comment (evalsocket) [reply selfservice/hook/session_destroyer_test.go]: done

File: session/persistence.go [hunk 13]
```diff
@@ -29,6 +29,9 @@ type Persister interface {
 
 	// Delete removes a session from the store
 	DeleteSession(ctx context.Context, sid uuid.UUID) error
+
+	// Delete removes a session the store
```
Comment (aeneasr) [inline session/persistence.go]: This should describe what the function is doing, so more something like:

```suggestion
	// DeleteSessionsFor removes all active session from the store for the given identity.
```

**Comment by aeneasr** [comment]
Awesome, thank you for your hard work! Well done :)
